lossy_compressors/vector_quantizer.py

```diff
@@ -0,0 +1,223 @@
+from typing import List
+from compressors.fixed_bitwidth_compressor import (
+    FixedBitwidthDecoder,
+    FixedBitwidthEncoder,
+    get_alphabet_fixed_bitwidth,
+)
+from core.data_block import DataBlock
+from core.data_encoder_decoder import DataDecoder, DataEncoder
+import numpy as np
+import random
+import scipy
+from utils.bitarray_utils import BitArray, bitarray_to_uint, uint_to_bitarray
+from external_compressors.pickle_external import PickleEncoder, PickleDecoder
+
+
+def mse(v1, v2):
+    d = np.linalg.norm(v1 - v2, ord=2)  # l2_norm
+    loss = d * d / (v1.size)  # avg l2 loss
+    return loss
+
+
+def find_nearest(codebook_npy, data_npy, dist_func):
+    """
+    codebook_npy -> [V,D] where V -> num vectors in the codebook, D -> dim of each vector
+    data_npy -> [D] sized vector
+    """
+    distances = [dist_func(c, data_npy) for c in codebook_npy]
+    min_ind = np.argmin(distances)
+    return min_ind, distances[min_ind]
+
+
+def build_kmeans_codebook_fast(data_npy, num_vectors, dist_thresh=1e-4, max_iter=100):
+    """
+    NOTE: implicitly assumes dist_func is mse
+    output codebook -> [num_vectors,dim]
+    data_npy -> [N/D, D] sizes, where N is the size of data
+    """
+    codebook, _ = scipy.cluster.vq.kmeans(
+        data_npy, num_vectors, iter=max_iter, thresh=dist_thresh, seed=0
+    )
+    return codebook
+
+
+def build_kmeans_codebook(
+    data_npy, num_vectors, distortion_func=mse, dist_thresh=1e-4, max_iter=100
+):
+    """
+    NOTE: slow, as we are doing exact nearest neighbor etc.
+    output codebook -> [num_vectors,dim]
+    data_npy -> [N/D, D] sizes, where N is the size of data
+    """
+
+    # define initial tensors to be samples from existing data (without replacement)
+    data_npy = [np.array(v) for v in data_npy.tolist()]
+    codebook = random.sample(data_npy, num_vectors)
+    codebook = np.stack(codebook, axis=0)
+
+    ## for the entire dataset, find the closest centroid
+    # repeat until convergence
+    avg_distortion_arr = []
+    for it in range(max_iter):
+        # define intermediate tensors and vals
+        counts = np.ones(num_vectors)
+        centroids = codebook.copy()  # current centroid
+        sum_distances = 0
+
+        # go over the input data and average
+        for data_vector in data_npy:
+            min_ind, dist = find_nearest(codebook, data_vector, distortion_func)
```

Comment:
@tpulkit  -> min is found wrt to the codebook, which is updated using the centroids only at the end